subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c

```diff
@@ -448,10 +469,107 @@ static int open(const struct device *instance)
 
 }
 
+static int get_tx_buffer_size(const struct device *instance, void *token)
+{
+	struct backend_data_t *data = instance->data;
+	struct ipc_rpmsg_instance *rpmsg_inst;
+	struct rpmsg_device *rdev;
+	int size;
+
+	rpmsg_inst = &data->rpmsg_inst;
+	rdev = rpmsg_virtio_get_rpmsg_device(&rpmsg_inst->rvdev);
+
+	size = rpmsg_virtio_get_buffer_size(rdev);
+	if (size < 0) {
+		return -EIO;
+	}
+
+	return size;
+}
+
+static int get_tx_buffer(const struct device *instance, void *token,
+			 void **r_data, uint32_t *size, k_timeout_t wait)
+{
+	struct ipc_rpmsg_ept *rpmsg_ept;
+	void *payload;
+	int buf_size;
+
+	rpmsg_ept = (struct ipc_rpmsg_ept *) token;
+
+	if (!r_data || !size) {
+		return -EINVAL;
+	}
+
+	/* OpenAMP only supports a binary wait / no-wait */
+	if (!K_TIMEOUT_EQ(wait, K_FOREVER) && !K_TIMEOUT_EQ(wait, K_NO_WAIT)) {
+		return -ENOTSUP;
+	}
+
+	/* The user requested a specific size */
+	if (*size) {
+		buf_size = get_tx_buffer_size(instance, token);
+		if (buf_size < 0) {
+			return -EIO;
+		}
+
+		/* Not the requested size */
+		if (*size != buf_size) {
+			*size = buf_size;
+			return -ENOMEM;
+		}
```

Comment:
I'm not sure if this implementation works correctly. Let's say I'm developing a module sending 12 bytes of data between two cores. I use an endpoint registered in ipc_service for this purpose. I'm calling `get_tx_buffer(..., 12, ...);` sequentially to allocate buffers for my message. But every time I call this function I receive `-ENOMEM` even though virtio buffers could easily hold my 12 bytes long message. When I try again, I receive `-ENOMEM`, and again, and again.

I think it should be more like `if (*size > buf_size) { return -ENOMEM; } else { /* allocated buffer is larger than requested, but it is fine */ ... return 0; }`

Reply:
ok, so you want to return `-ENOMEM` only when the requested size is bigger than the buffer size.

Reply:
I'm thinking if there is any consequence in releasing buffer procedures.

In tx buffer releasing scenario:
```
uint32_t len = 16;
void *data;
r = ipc_service_get_tx_buffer(ept, &data, &len, NO_WAIT); // Allocated buffer is 256 bytes long, function returns 0
...
r = ipc_service_drop_tx_buffer(ept, data);
```
There is no buffer size parameter passed to `drop_tx_buffer()` so probably there is no problem.

In sending tx buffer scenario:
```
uint32_t len = 16;
void *data;
r = ipc_service_get_tx_buffer(ept, &data, &len, NO_WAIT); // Allocated buffer is 256 bytes long, function returns 0
...
r = ipc_service_send_nocopy(ept, data, 16);
```
Allocated buffer is 256 bytes long, but `send_nocopy()` function uses only 16 bytes out of this buffer. I think it is fine to assume that `send_nocopy()` will release whole 256 bytes here anyway.

What do you think?

Reply:
Then yes, I would prefer `-ENOMEM` only when available size is smaller than requested. If allocated buffer is bigger `0` seems better.